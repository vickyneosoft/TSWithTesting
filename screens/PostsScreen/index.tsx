import { NativeStackScreenProps } from '@react-navigation/native-stack';
import React, { useCallback, useEffect, useMemo } from 'react';
import {
  ActivityIndicator,
  FlatList,
  InteractionManager,
  StyleSheet,
  View,
} from 'react-native';
import BoldText from '../../components/BoldText';
import PostComponent from '../../components/PostComponent';
import colors from '../../constants/colors';
import usePosts from '../../hooks/usePosts';
import { Post } from '../../types';
import { keyExtractorHandler } from '../../utils/miscUtils';

const PostsScreen = (props: NativeStackScreenProps<any, any>) => {
  const { navigation } = props;
  const { error, isLoading, posts, fetchPosts } = usePosts();

  useEffect(() => {
    InteractionManager.runAfterInteractions(fetchPosts)
  }, [fetchPosts]);

  const onPostItemPressHandler = useCallback(
    (selectedPostId: number) => {
      navigation.navigate('postDetails', {
        postId: selectedPostId,
      });
    },
    [navigation],
  );

  const renderItemHandler = useCallback(
    ({ item }: { item: Post }) => {
      try {
        const { title, id } = item;
        return (
          <PostComponent
            title={title}
            onPress={onPostItemPressHandler.bind(null, id)} />
        );
      } catch (err: any) {
        console.log('Error : ', err.message);
        return null;
      }
    },
    [onPostItemPressHandler],
  );

  const listEmptyComponent = useMemo(() => {
    if (error) {
      return (
        <View style={styles.noDataFoundContainer}>
          <BoldText>{error}</BoldText>
        </View>
      );
    } else if (isLoading) {
      return (
        <View style={styles.noDataFoundContainer}>
          <ActivityIndicator size={'large'} color={colors.grey} />
        </View>
      );
    } else {
      return (
        <View style={styles.noDataFoundContainer}>
          <BoldText>{'No data found!'}</BoldText>
        </View>
      );
    }
  }, [isLoading, error]);

  return (
    <FlatList
      testID='lstPosts'
      data={posts}
      keyExtractor={keyExtractorHandler}
      renderItem={renderItemHandler}
      contentContainerStyle={styles.container}
      ListEmptyComponent={listEmptyComponent}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    minHeight: '100%',
    paddingBottom: 20,
    backgroundColor: colors.grey,
  },
  noDataFoundContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default PostsScreen;
