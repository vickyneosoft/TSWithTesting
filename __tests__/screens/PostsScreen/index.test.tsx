import React from 'react';
import renderer, { act, ReactTestInstance, ReactTestRenderer, ReactTestRendererJSON } from 'react-test-renderer'
import PostsScreen from '../../../screens/PostsScreen';

const navigation = {
    navigate: (screenName: string, params: any) => { }
}

var screen: ReactTestRenderer, tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null

describe('<PostsScreen /> rendering', () => {
    it('PostsScreen screen renders correctly', async () => {
        await act(async () => {
            screen = await renderer.create(
                <PostsScreen navigation={navigation} />
            )
        })
        tree = screen.toJSON();
        expect(tree).toMatchSnapshot();
    });
});