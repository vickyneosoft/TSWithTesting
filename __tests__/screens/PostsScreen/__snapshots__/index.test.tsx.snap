// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostsScreen /> rendering PostsScreen screen renders correctly 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <BoldText>
        No data found!
      </BoldText>
    </View>
  }
  contentContainerStyle={
    {
      "backgroundColor": "#343434",
      "minHeight": "100%",
      "paddingBottom": 20,
    }
  }
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  testID="lstPosts"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          {
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        No data found!
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
